File: platform/android/src/org/fathens/cordova/plugin/fabric/FabricAnswers.java
package org.fathens.cordova.plugin.fabric;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.math.BigDecimal;
import java.util.Currency;
import java.util.Iterator;

import org.apache.cordova.CallbackContext;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import com.crashlytics.android.answers.AddToCartEvent;
import com.crashlytics.android.answers.Answers;
import com.crashlytics.android.answers.AnswersEvent;
import com.crashlytics.android.answers.ContentViewEvent;
import com.crashlytics.android.answers.CustomEvent;
import com.crashlytics.android.answers.InviteEvent;
import com.crashlytics.android.answers.LevelEndEvent;
import com.crashlytics.android.answers.LevelStartEvent;
import com.crashlytics.android.answers.LoginEvent;
import com.crashlytics.android.answers.PurchaseEvent;
import com.crashlytics.android.answers.RatingEvent;
import com.crashlytics.android.answers.SearchEvent;
import com.crashlytics.android.answers.ShareEvent;
import com.crashlytics.android.answers.SignUpEvent;
import com.crashlytics.android.answers.StartCheckoutEvent;

public class FabricAnswers {
	static public Runnable execute(final String action, final JSONArray data, final CallbackContext callbackContext)
			throws JSONException {
		try {
			final Class<FabricAnswers> obj = FabricAnswers.class;
			final Method method = obj.getMethod(action, JSONArray.class, CallbackContext.class);
			return (Runnable) method.invoke(obj, data, callbackContext);
		} catch (NoSuchMethodException e) {
			return null;
		} catch (IllegalAccessException ex) {
			throw new RuntimeException(ex);
		} catch (IllegalArgumentException ex) {
			throw new RuntimeException(ex);
		} catch (InvocationTargetException ex) {
			throw new RuntimeException(ex);
		}
	}

	static public void putCustom(AnswersEvent<?> event, JSONObject custom) throws JSONException {
		for (@SuppressWarnings("unchecked")
		Iterator<String> iterator = custom.keys(); iterator.hasNext();) {
			final String key = iterator.next();
			final Object value = custom.get(key);
			if (value instanceof String)
				event.putCustomAttribute(key, (String) value);
			if (value instanceof Number)
				event.putCustomAttribute(key, (Number) value);
		}
	}

	static public Runnable eventPurchase(final JSONArray args, final CallbackContext context) throws JSONException {
		final JSONObject obj = args.getJSONObject(0);
		return new Runnable() {
			public void run() {
				try {
					final PurchaseEvent event = new PurchaseEvent();
					if (obj.has("itemPrice"))
						event.putItemPrice(new BigDecimal(obj.getInt("itemPrice")));
					if (obj.has("currency"))
						event.putCurrency(Currency.getInstance(obj.getString("currency")));
					if (obj.has("itemName"))
						event.putItemName(obj.getString("itemName"));
					if (obj.has("itemType"))
						event.putItemType(obj.getString("itemType"));
					if (obj.has("itemId"))
						event.putItemId(obj.getString("itemId"));
					if (obj.has("success"))
						event.putSuccess(obj.getBoolean("success"));
					if (obj.has("custom")) {
						putCustom(event, obj.getJSONObject("custom"));
					}
					Answers.getInstance().logPurchase(event);
					context.success();
				} catch (Exception ex) {
					throw new RuntimeException(ex);
				}
			}
		};
	}

	static public Runnable eventAddToCart(final JSONArray args, final CallbackContext context) throws JSONException {
		final JSONObject obj = args.getJSONObject(0);
		return new Runnable() {
			public void run() {
				try {
					final AddToCartEvent event = new AddToCartEvent();
					if (obj.has("itemPrice"))
						event.putItemPrice(new BigDecimal(obj.getInt("itemPrice")));
					if (obj.has("currency"))
						event.putCurrency(Currency.getInstance(obj.getString("currency")));
					if (obj.has("itemName"))
						event.putItemName(obj.getString("itemName"));
					if (obj.has("itemType"))
						event.putItemType(obj.getString("itemType"));
					if (obj.has("itemId"))
						event.putItemId(obj.getString("itemId"));
					if (obj.has("custom")) {
						putCustom(event, obj.getJSONObject("custom"));
					}
					Answers.getInstance().logAddToCart(event);
					context.success();
				} catch (Exception ex) {
					throw new RuntimeException(ex);
				}
			}
		};
	}

	static public Runnable eventStartCheckout(final JSONArray args, final CallbackContext context) throws JSONException {
		final JSONObject obj = args.getJSONObject(0);
		return new Runnable() {
			public void run() {
				try {
					final StartCheckoutEvent event = new StartCheckoutEvent();
					if (obj.has("totalPrice"))
						event.putTotalPrice(new BigDecimal(obj.getInt("totalPrice")));
					if (obj.has("currency"))
						event.putCurrency(Currency.getInstance(obj.getString("currency")));
					if (obj.has("itemCount"))
						event.putItemCount(obj.getInt("itemCount"));
					if (obj.has("custom")) {
						putCustom(event, obj.getJSONObject("custom"));
					}
					Answers.getInstance().logStartCheckout(event);
					context.success();
				} catch (Exception ex) {
					throw new RuntimeException(ex);
				}
			}
		};
	}

	static public Runnable eventContentView(final JSONArray args, final CallbackContext context) throws JSONException {
		final JSONObject obj = args.getJSONObject(0);
		return new Runnable() {
			public void run() {
				try {
					final ContentViewEvent event = new ContentViewEvent();
					if (obj.has("contentName"))
						event.putContentName(obj.getString("contentName"));
					if (obj.has("contentType"))
						event.putContentType(obj.getString("contentType"));
					if (obj.has("contentId"))
						event.putContentId(obj.getString("contentId"));
					if (obj.has("custom")) {
						putCustom(event, obj.getJSONObject("custom"));
					}
					Answers.getInstance().logContentView(event);
					context.success();
				} catch (Exception ex) {
					throw new RuntimeException(ex);
				}
			}
		};
	}

	static public Runnable eventSearch(final JSONArray args, final CallbackContext context) throws JSONException {
		final JSONObject obj = args.getJSONObject(0);
		return new Runnable() {
			public void run() {
				try {
					final SearchEvent event = new SearchEvent();
					if (obj.has("query"))
						event.putQuery(obj.getString("query"));
					if (obj.has("custom")) {
						putCustom(event, obj.getJSONObject("custom"));
					}
					Answers.getInstance().logSearch(event);
					context.success();
				} catch (Exception ex) {
					throw new RuntimeException(ex);
				}
			}
		};
	}

	static public Runnable eventShare(final JSONArray args, final CallbackContext context) throws JSONException {
		final JSONObject obj = args.getJSONObject(0);
		return new Runnable() {
			public void run() {
				try {
					final ShareEvent event = new ShareEvent();
					if (obj.has("method"))
						event.putMethod(obj.getString("method"));
					if (obj.has("contentName"))
						event.putContentName(obj.getString("contentName"));
					if (obj.has("contentType"))
						event.putContentType(obj.getString("contentType"));
					if (obj.has("contentId"))
						event.putContentId(obj.getString("contentId"));
					if (obj.has("custom")) {
						putCustom(event, obj.getJSONObject("custom"));
					}
					Answers.getInstance().logShare(event);
					context.success();
				} catch (Exception ex) {
					throw new RuntimeException(ex);
				}
			}
		};
	}

	static public Runnable eventRating(final JSONArray args, final CallbackContext context) throws JSONException {
		final JSONObject obj = args.getJSONObject(0);
		return new Runnable() {
			public void run() {
				try {
					final RatingEvent event = new RatingEvent();
					if (obj.has("rating"))
						event.putRating(obj.getInt("rating"));
					if (obj.has("contentName"))
						event.putContentName(obj.getString("contentName"));
					if (obj.has("contentType"))
						event.putContentType(obj.getString("contentType"));
					if (obj.has("contentId"))
						event.putContentId(obj.getString("contentId"));
					if (obj.has("custom")) {
						putCustom(event, obj.getJSONObject("custom"));
					}
					Answers.getInstance().logRating(event);
					context.success();
				} catch (Exception ex) {
					throw new RuntimeException(ex);
				}
			}
		};
	}

	static public Runnable eventSignUp(final JSONArray args, final CallbackContext context) throws JSONException {
		final JSONObject obj = args.getJSONObject(0);
		return new Runnable() {
			public void run() {
				try {
					final SignUpEvent event = new SignUpEvent();
					if (obj.has("method"))
						event.putMethod(obj.getString("method"));
					if (obj.has("success"))
						event.putSuccess(obj.getBoolean("success"));
					if (obj.has("custom")) {
						putCustom(event, obj.getJSONObject("custom"));
					}
					Answers.getInstance().logSignUp(event);
					context.success();
				} catch (Exception ex) {
					throw new RuntimeException(ex);
				}
			}
		};
	}

	static public Runnable eventLogin(final JSONArray args, final CallbackContext context) throws JSONException {
		final JSONObject obj = args.getJSONObject(0);
		return new Runnable() {
			public void run() {
				try {
					final LoginEvent event = new LoginEvent();
					if (obj.has("method"))
						event.putMethod(obj.getString("method"));
					if (obj.has("success"))
						event.putSuccess(obj.getBoolean("success"));
					if (obj.has("custom")) {
						putCustom(event, obj.getJSONObject("custom"));
					}
					Answers.getInstance().logLogin(event);
					context.success();
				} catch (Exception ex) {
					throw new RuntimeException(ex);
				}
			}
		};
	}

	static public Runnable eventInvite(final JSONArray args, final CallbackContext context) throws JSONException {
		final JSONObject obj = args.getJSONObject(0);
		return new Runnable() {
			public void run() {
				try {
					final InviteEvent event = new InviteEvent();
					if (obj.has("method"))
						event.putMethod(obj.getString("method"));
					if (obj.has("custom")) {
						putCustom(event, obj.getJSONObject("custom"));
					}
					Answers.getInstance().logInvite(event);
					context.success();
				} catch (Exception ex) {
					throw new RuntimeException(ex);
				}
			}
		};
	}

	static public Runnable eventLevelStart(final JSONArray args, final CallbackContext context) throws JSONException {
		final JSONObject obj = args.getJSONObject(0);
		return new Runnable() {
			public void run() {
				try {
					final LevelStartEvent event = new LevelStartEvent();
					if (obj.has("levelName"))
						event.putLevelName(obj.getString("levelName"));
					if (obj.has("custom")) {
						putCustom(event, obj.getJSONObject("custom"));
					}
					Answers.getInstance().logLevelStart(event);
					context.success();
				} catch (Exception ex) {
					throw new RuntimeException(ex);
				}
			}
		};
	}

	static public Runnable eventLevelEnd(final JSONArray args, final CallbackContext context) throws JSONException {
		final JSONObject obj = args.getJSONObject(0);
		return new Runnable() {
			public void run() {
				try {
					final LevelEndEvent event = new LevelEndEvent();
					if (obj.has("levelName"))
						event.putLevelName(obj.getString("levelName"));
					if (obj.has("score"))
						event.putScore(obj.getInt("score"));
					if (obj.has("success"))
						event.putSuccess(obj.getBoolean("success"));
					if (obj.has("custom")) {
						putCustom(event, obj.getJSONObject("custom"));
					}
					Answers.getInstance().logLevelEnd(event);
					context.success();
				} catch (Exception ex) {
					throw new RuntimeException(ex);
				}
			}
		};
	}

	static public Runnable eventCustom(final JSONArray args, final CallbackContext context) throws JSONException {
		final JSONObject obj = args.getJSONObject(0);
		return new Runnable() {
			public void run() {
				try {
					final CustomEvent event = new CustomEvent(obj.getString("name"));
					if (obj.has("attributes")) {
						putCustom(event, obj.getJSONObject("attributes"));
					}
					Answers.getInstance().logCustom(event);
					context.success();
				} catch (Exception ex) {
					throw new RuntimeException(ex);
				}
			}
		};
	}
}

File: platforms/android/src/main/kotlin/FabricAnswers.kt
package org.fathens.cordova.plugin.fabric

import java.math.BigDecimal
import java.util.Currency
import org.apache.cordova.CallbackContext
import org.apache.cordova.CordovaPlugin
import org.json.JSONArray
import org.json.JSONObject
import com.crashlytics.android.answers.*

class ParamDict(val obj: JSONObject) {
    fun opt(name: String): Any? = obj.opt(name)

    fun getJSONObject(name: String): JSONObject? = opt(name) as? JSONObject
    fun getString(name: String): String? = opt(name) as? String
    fun getBoolean(name: String): Boolean? = opt(name) as? Boolean
    fun getNumber(name: String): Number? = opt(name) as? Number
    fun getDecimal(name: String): BigDecimal? = getNumber(name)?.let { BigDecimal(it.toDouble()) }
    fun getCurrency(name: String): Currency? = getString(name)?.let { Currency.getInstance(it)}
}

public class FabricAnswers : CordovaPlugin() {
    override fun execute(action: String, data: JSONArray, callbackContext: CallbackContext): Boolean {
        try {
            val method = javaClass.getMethod(action, data.javaClass)
            val runner = method.invoke(this, data) as (() -> Unit)
            cordova.getThreadPool().execute {
                runner()
                callbackContext.success()
            }
            return true
        } catch (e: NoSuchMethodException) {
            return false
        }
    }

    public fun putCustom(event: AnswersEvent<*>, custom: JSONObject) {
        val iterator = custom.keys()
        while (iterator.hasNext()) {
            val key = iterator.next()
            val value = custom.get(key)
            when (value) {
                is String -> event.putCustomAttribute(key, value)
                is Number -> event.putCustomAttribute(key, value)
                else -> event.putCustomAttribute(key, "$value")
            }
        }
    }

    public fun eventPurchase(args: JSONArray): (() -> Unit) {
        val obj = if (args.length() > 0) args.getJSONObject(0) else null
        return {
            val event = PurchaseEvent()
            if (obj != null) {
                val dict = ParamDict(obj)
                dict.getDecimal("itemPrice")?.let { event.putItemPrice(it) }
                dict.getCurrency("currency")?.let { event.putCurrency(it) }
                dict.getString("itemName")?.let { event.putItemName(it) }
                dict.getString("itemType")?.let { event.putItemType(it) }
                dict.getString("itemId")?.let { event.putItemId(it) }
                dict.getBoolean("success")?.let { event.putSuccess(it) }
                dict.getJSONObject("custom")?.let { putCustom(event, it) }
            }
            Answers.getInstance().logPurchase(event)
        }
    }

    public fun eventAddToCart(args: JSONArray): (() -> Unit) {
        val obj = if (args.length() > 0) args.getJSONObject(0) else null
        return {
            val event = AddToCartEvent()
            if (obj != null) {
                val dict = ParamDict(obj)
                dict.getDecimal("itemPrice")?.let { event.putItemPrice(it) }
                dict.getCurrency("currency")?.let { event.putCurrency(it) }
                dict.getString("itemName")?.let { event.putItemName(it) }
                dict.getString("itemType")?.let { event.putItemType(it) }
                dict.getString("itemId")?.let { event.putItemId(it) }
                dict.getJSONObject("custom")?.let { putCustom(event, it) }
            }
            Answers.getInstance().logAddToCart(event)
        }
    }

    public fun eventStartCheckout(args: JSONArray): (() -> Unit) {
        val obj = if (args.length() > 0) args.getJSONObject(0) else null
        return {
            val event = StartCheckoutEvent()
            if (obj != null) {
                val dict = ParamDict(obj)
                dict.getDecimal("totalPrice")?.let { event.putTotalPrice(it) }
                dict.getCurrency("currency")?.let { event.putCurrency(it) }
                dict.getNumber("itemCount")?.let { event.putItemCount(it.toInt()) }
                dict.getJSONObject("custom")?.let { putCustom(event, it) }
            }
            Answers.getInstance().logStartCheckout(event)
        }
    }

    public fun eventContentView(args: JSONArray): (() -> Unit) {
        val obj = if (args.length() > 0) args.getJSONObject(0) else null
        return {
            val event = ContentViewEvent()
            if (obj != null) {
                val dict = ParamDict(obj)
                dict.getString("contentName")?.let { event.putContentName(it) }
                dict.getString("contentType")?.let { event.putContentType(it) }
                dict.getString("contentId")?.let { event.putContentId(it) }
                dict.getJSONObject("custom")?.let { putCustom(event, it) }
            }
            Answers.getInstance().logContentView(event)
        }
    }

    public fun eventSearch(args: JSONArray): (() -> Unit) {
        val obj = if (args.length() > 0) args.getJSONObject(0) else null
        return {
            val event = SearchEvent()
            if (obj != null) {
                val dict = ParamDict(obj)
                dict.getString("query")?.let { event.putQuery(it) }
                dict.getJSONObject("custom")?.let { putCustom(event, it) }
            }
            Answers.getInstance().logSearch(event)
        }
    }

    public fun eventShare(args: JSONArray): (() -> Unit) {
        val obj = if (args.length() > 0) args.getJSONObject(0) else null
        return {
            val event = ShareEvent()
            if (obj != null) {
                val dict = ParamDict(obj)
                dict.getString("method")?.let { event.putMethod(it) }
                dict.getString("contentName")?.let { event.putContentName(it) }
                dict.getString("contentType")?.let { event.putContentType(it) }
                dict.getString("contentId")?.let { event.putContentId(it) }
                dict.getJSONObject("custom")?.let { putCustom(event, it) }
            }
            Answers.getInstance().logShare(event)
        }
    }

    public fun eventRating(args: JSONArray): (() -> Unit) {
        val obj = if (args.length() > 0) args.getJSONObject(0) else null
        return {
            val event = RatingEvent()
            if (obj != null) {
                val dict = ParamDict(obj)
                dict.getNumber("rating")?.let { event.putRating(it.toInt()) }
                dict.getString("contentName")?.let { event.putContentName(it) }
                dict.getString("contentType")?.let { event.putContentType(it) }
                dict.getString("contentId")?.let { event.putContentId(it) }
                dict.getJSONObject("custom")?.let { putCustom(event, it) }
            }
            Answers.getInstance().logRating(event)
        }
    }

    public fun eventSignUp(args: JSONArray): (() -> Unit) {
        val obj = if (args.length() > 0) args.getJSONObject(0) else null
        return {
            val event = SignUpEvent()
            if (obj != null) {
                val dict = ParamDict(obj)
                dict.getString("method")?.let { event.putMethod(it) }
                dict.getBoolean("success")?.let { event.putSuccess(it) }
                dict.getJSONObject("custom")?.let { putCustom(event, it) }
            }
            Answers.getInstance().logSignUp(event)
        }
    }

    public fun eventLogin(args: JSONArray): (() -> Unit) {
        val obj = if (args.length() > 0) args.getJSONObject(0) else null
        return {
            val event = LoginEvent()
            if (obj != null) {
                val dict = ParamDict(obj)
                dict.getString("method")?.let { event.putMethod(it) }
                dict.getBoolean("success")?.let { event.putSuccess(it) }
                dict.getJSONObject("custom")?.let { putCustom(event, it) }
            }
            Answers.getInstance().logLogin(event)
        }
    }

    public fun eventInvite(args: JSONArray): (() -> Unit) {
        val obj = if (args.length() > 0) args.getJSONObject(0) else null
        return {
            val event = InviteEvent()
            if (obj != null) {
                val dict = ParamDict(obj)
                dict.getString("method")?.let { event.putMethod(it) }
                dict.getJSONObject("custom")?.let { putCustom(event, it) }
            }
            Answers.getInstance().logInvite(event)
        }
    }

    public fun eventLevelStart(args: JSONArray): (() -> Unit) {
        val obj = if (args.length() > 0) args.getJSONObject(0) else null
        return {
            val event = LevelStartEvent()
            if (obj != null) {
                val dict = ParamDict(obj)
                dict.getString("levelName")?.let { event.putLevelName(it) }
                dict.getJSONObject("custom")?.let { putCustom(event, it) }
            }
            Answers.getInstance().logLevelStart(event)
        }
    }

    public fun eventLevelEnd(args: JSONArray): (() -> Unit) {
        val obj = if (args.length() > 0) args.getJSONObject(0) else null
        return {
            val event = LevelEndEvent()
            if (obj != null) {
                val dict = ParamDict(obj)
                dict.getString("levelName")?.let { event.putLevelName(it) }
                dict.getNumber("score")?.let { event.putScore(it) }
                dict.getBoolean("success")?.let { event.putSuccess(it) }
                dict.getJSONObject("custom")?.let { putCustom(event, it) }
            }
            Answers.getInstance().logLevelEnd(event)
        }
    }

    public fun eventCustom(args: JSONArray): (() -> Unit) {
        val obj = if (args.length() > 0) args.getJSONObject(0) else null
        return {
            val event = CustomEvent(obj?.let { ParamDict(it) }?.getString("name") ?: "NoName")
            if (obj != null) {
                val dict = ParamDict(obj)
                dict.getJSONObject("attributes")?.let { putCustom(event, it) }
            }
            Answers.getInstance().logCustom(event)
        }
    }
}
